media/starboard/starboard_cdm.cc

```diff
@@ -12,371 +12,324 @@
+  auto it = session_map_.find(session_id);
+  if (it == session_map_.end()) {
+    LOG(WARNING) << "Update session - session doesn't exist";
+    promise->reject(CdmPromise::Exception::INVALID_STATE_ERROR, 0,
+                    "Session doesn't exist.");
+    return;
+  }
+
+  // Caller should NOT pass in an empty response.
+  DCHECK(!response.empty());
+
+  uint32_t promise_id = promises_.SavePromise(std::move(promise));
+
+  StarboardCdmSession* session = it->second.get();
+  DCHECK(session);
 
   // Store the context of the call.
   SessionUpdate session_update;
-  session_update.updated_callback = session_updated_callback;
-  session_update.did_not_update_callback = session_did_not_update_callback;
+  session_update.promise_id = promise_id;
+  session_update.session = session;
   int ticket = next_ticket_++;
 
   if (!SbDrmTicketIsValid(ticket)) {
-    LOG(ERROR) << "Update session with invalid ticket";
+    LOG(WARNING) << "Update session with invalid ticket";
     return;
   }
 
   ticket_to_session_update_map_.insert(std::make_pair(ticket, session_update));
 
-  LOG(INFO) << "Update session of drm system (" << wrapped_drm_system_
-            << "), key length: " << key_length << ", ticket: " << ticket
+  LOG(INFO) << "Update session of drm system (" << sb_drm_
+            << "), key length: " << response.size() << ", ticket: " << ticket
             << ", session id: " << session_id;
 
-  media_metrics_provider_.StartTrackingAction(
-      MediaAction::SBDRM_UPDATE_SESSION);
-  SbDrmUpdateSession(wrapped_drm_system_, ticket, key, key_length,
+  SbDrmUpdateSession(sb_drm_, ticket, response.data(), response.size(),
                      session_id.c_str(), session_id.size());
-  media_metrics_provider_.EndTrackingAction(MediaAction::SBDRM_UPDATE_SESSION);
 }
 
-void DrmSystem::CloseSession(const std::string& session_id) {
+// Runs the parallel steps from https://w3c.github.io/encrypted-media/#close.
+void StarboardCdm::CloseSession(const std::string& session_id,
+                                std::unique_ptr<SimpleCdmPromise> promise) {
   DCHECK(task_runner_->RunsTasksInCurrentSequence());
-  DCHECK(is_valid());
+  LOG(INFO) << "StarboardCdm - Close session";
+  auto it = session_map_.find(session_id);
+  if (it == session_map_.end()) {
+    promise->reject(CdmPromise::Exception::INVALID_STATE_ERROR, 0,
+                    "Session doesn't exist.");
+    return;
+  }
 
-  LOG(INFO) << "Close session of drm system (" << wrapped_drm_system_
-            << "), session id: " << session_id;
+  StarboardCdmSession* session = it->second.get();
+  DCHECK(session);
 
-  media_metrics_provider_.StartTrackingAction(MediaAction::SBDRM_CLOSE_SESSION);
-  SbDrmCloseSession(wrapped_drm_system_, session_id.c_str(), session_id.size());
-  media_metrics_provider_.EndTrackingAction(MediaAction::SBDRM_CLOSE_SESSION);
+  SbDrmCloseSession(sb_drm_, session_id.c_str(), session_id.size());
+
+  session_map_.erase(session_id);
```

Comment:
We should carefully consider how to address the removed session id from other maps.  For example, we may need to ignore an in progress update request when the session is closed.  Feel free to leave this as a TODO.

Reply:
done